src/ess/reduce/time_of_flight/toa_to_tof.py

```diff
@@ -0,0 +1,520 @@
+        values=np.arange(
+            min_dist.value, np.nextafter(max_dist.value, np.inf), res.value
+        ),
+        unit=distance_unit,
+    )
+    distances = sc.midpoints(dist_edges)
+
+    time_unit = simulation.time_of_arrival.unit
+    toas = simulation.time_of_arrival + (distances / simulation.speed).to(
+        unit=time_unit, copy=False
+    )
+
+    # Compute time-of-flight for all neutrons
+    wavs = sc.broadcast(simulation.wavelength.to(unit="m"), sizes=toas.sizes).flatten(
+        to="event"
+    )
+    dist = sc.broadcast(distances + simulation_distance, sizes=toas.sizes).flatten(
+        to="event"
+    )
+    tofs = dist * sc.constants.m_n
+    tofs *= wavs
+    tofs /= sc.constants.h
+
+    data = sc.DataArray(
+        data=sc.broadcast(simulation.weight, sizes=toas.sizes).flatten(to="event"),
+        coords={
+            "toa": toas.flatten(to="event"),
+            "tof": tofs.to(unit=time_unit, copy=False),
+            "distance": dist,
+        },
+    )
+
+    binned = data.bin(distance=dist_edges + simulation_distance, toa=toa_resolution)
+    # Weighted mean of tof inside each bin
+    mean_tof = (
+        binned.bins.data * binned.bins.coords["tof"]
+    ).bins.sum() / binned.bins.sum()
+    # Compute the variance of the tofs to track regions with large uncertainty
+    variance = (
+        binned.bins.data * (binned.bins.coords["tof"] - mean_tof) ** 2
+    ).bins.sum() / binned.bins.sum()
+
+    mean_tof.variances = variance.values
+
+    # Convert coordinates to midpoints
+    mean_tof.coords["toa"] = sc.midpoints(mean_tof.coords["toa"])
+    mean_tof.coords["distance"] = sc.midpoints(mean_tof.coords["distance"])
+
+    return TimeOfFlightLookupTable(mean_tof)
+
+
+def masked_tof_lookup_table(
+    tof_lookup: TimeOfFlightLookupTable,
+    error_threshold: LookupTableRelativeErrorThreshold,
+) -> MaskedTimeOfFlightLookupTable:
+    """
+    Mask regions of the lookup table where the variance of the projected time-of-flight
+    is larger than a given threshold.
+
+    Parameters
+    ----------
+    tof_lookup:
+        Lookup table giving time-of-flight as a function of distance and
+        time-of-arrival.
+    variance_threshold:
+        Threshold for the variance of the projected time-of-flight above which regions
+        are masked.
+    """
+    relative_error = sc.stddevs(tof_lookup.data) / sc.values(tof_lookup.data)
+    mask = relative_error > sc.scalar(error_threshold)
+    out = tof_lookup.copy()
+    # Use numpy for indexing as table is 2D
+    out.values[mask.values] = np.nan
+    return MaskedTimeOfFlightLookupTable(out)
+
+
+def find_fastest_neutron(simulation: SimulationResults) -> FastestNeutron:
+    """
+    Find the fastest neutron in the simulation results.
+    """
+    ind = np.argmax(simulation.speed.values)
+    return FastestNeutron(
+        time_of_arrival=simulation.time_of_arrival[ind],
+        speed=simulation.speed[ind],
+        distance=simulation.distance,
```

Comment:
No index? The `SimulationResults` docstring is not very telling.

Reply:
No index because the distance is a single number.